File: packages/oidc_resource_server/test/claims_test.dart
import 'package:oidc_resource_server/oidc_resource_server.dart';
import 'package:test/test.dart';

void main() {
  group('matchesFgacGrant', () {
    test('exact resource id', () {
      final rels = [
        FgacRelationClaim(resourceType: 'blog_post', resourceId: 'a1', relation: 'editor'),
      ];
      expect(matchesFgacGrant(rels, 'blog_post', 'a1', relation: 'editor'), isTrue);
      expect(matchesFgacGrant(rels, 'blog_post', 'a2', relation: 'editor'), isFalse);
    });

    test('wildcard resource id', () {
      final rels = [
        FgacRelationClaim(resourceType: 'blog_post', resourceId: '*', relation: 'editor'),
      ];
      expect(matchesFgacGrant(rels, 'blog_post', 'any-id', relation: 'editor'), isTrue);
      expect(matchesFgacGrant(rels, 'blog_post', 'any-id', relation: 'viewer'), isFalse);
    });
  });

  group('effectivePermissionsOnResource', () {
    test('inherits viewer permissions from editor', () {
      const schema = <String, Map<String, FgacRelationDefinition>>{
        'doc': {
          'viewer': FgacRelationDefinition(permissions: {'read'}),
          'editor': FgacRelationDefinition(
            permissions: {'write'},
            inherits: ['viewer'],
          ),
        },
      };
      final perms = effectivePermissionsOnResource(
        schema: schema,
        resourceType: 'doc',
        relationNamesHeld: {'editor'},
      );
      expect(perms, containsAll(['read', 'write']));
    });
  });

  group('satisfiesRequirement', () {
    test('RequireAnyPermission', () {
      const schema = <String, Map<String, FgacRelationDefinition>>{
        'doc': {
          'viewer': FgacRelationDefinition(permissions: {'read'}),
        },
      };
      final claims = [
        FgacRelationClaim(resourceType: 'doc', resourceId: 'x', relation: 'viewer'),
      ];
      final ok = satisfiesRequirement(
        claims,
        RequireAnyPermission({'read'}, FgacResourceRef(type: 'doc', id: 'x'), schema),
      );
      expect(ok, isTrue);
    });
  });
}

File: packages/ts/oidc-resource-server/test/claims.test.ts
import { describe, expect, it } from 'vitest';

import { hasFgacPermission, matchesFgacGrant, parseFgacPermissions, parseFgacRelations } from '../src/claims';

describe('parseFgacRelations', () => {
	it('parses snake_case tuples', () => {
		const rels = parseFgacRelations({
			fgac_relations: [
				{ resource_type: 'blog', resource_id: '1', relation: 'admin' },
				{ resource_type: 'x', resource_id: 'y', relation: 'z' },
			],
		});
		expect(rels).toHaveLength(2);
		expect(matchesFgacGrant(rels, 'blog', '1', 'admin')).toBe(true);
		expect(matchesFgacGrant(rels, 'blog', '2', 'admin')).toBe(false);
		expect(matchesFgacGrant(rels, 'blog', 'any', 'admin')).toBe(false);
	});

	it('wildcard resource_id', () => {
		const rels = parseFgacRelations({
			fgac_relations: [{ resource_type: 'doc', resource_id: '*', relation: 'viewer' }],
		});
		expect(matchesFgacGrant(rels, 'doc', 'anything', 'viewer')).toBe(true);
	});
});

describe('parseFgacPermissions', () => {
	it('parses permission entries and de-duplicates permissions', () => {
		const perms = parseFgacPermissions({
			fgac_permissions: [
				{ resource_type: 'blog', resource_id: '1', permissions: ['read', 'write', 'read'] },
				{ resource_type: 'doc', resource_id: '*', permissions: ['read'] },
			],
		});
		expect(perms).toHaveLength(2);
		expect(hasFgacPermission(perms, 'blog', '1', 'write')).toBe(true);
		expect(hasFgacPermission(perms, 'blog', '1', 'delete')).toBe(false);
		expect(hasFgacPermission(perms, 'doc', 'anything', 'read')).toBe(true);
	});
});
